# Οδηγείες
#### 1.pip install -r req.txt
#### 2.στο ίδιο directory  στο data/companydata.csv πρέπει να υπάρχει το training set 
#### 3.στο ίδιο directory στο toEvaluate/test_unlabeled.csv πρέπει να υπάρχει το αρχείο που μας ενδιαφέρει
#### 5.αρχεία : results.txt το ζητούμενο , ScalersResult.txt και BestScalers.txt υπάρχουν για την αποφυγεί εκτέλεσεις χρονοβόλων γραμμών κώδικα

Import the neccessary libraries

In [6]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats.stats import pearsonr 
from scipy.stats.stats import spearmanr
from sklearn.metrics import brier_score_loss
from matplotlib import cm
from sklearn.preprocessing import Imputer
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix
from matplotlib.colors import ListedColormap
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import minmax_scale
from sklearn.preprocessing import MaxAbsScaler
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import RobustScaler
from sklearn.preprocessing import Normalizer
from sklearn.preprocessing.data import QuantileTransformer
from sklearn.datasets import make_moons, make_circles, make_classification
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

%matplotlib inline

# Όλοι οι classifiers που θα δοκιμάσουμε

In [7]:
names = ["NeuralNet", "Gaussian Process",
         "Decision Tree", "Random Forest", "Nearest Neighbors", "AdaBoost",
         "Naive Bayes", "QDA","Linear SVM", "RBF SVM"]

classifiers = [
    MLPClassifier(alpha=1),
    GaussianProcessClassifier(1.0 * RBF(1.0)),
    DecisionTreeClassifier(max_depth=5),
    RandomForestClassifier(max_depth=5, n_estimators=10, max_features=1),
    KNeighborsClassifier(3),
    AdaBoostClassifier(),
    GaussianNB(),
    QuadraticDiscriminantAnalysis(),
    SVC(kernel="linear", C=0.025),
    SVC(gamma=2, C=1)]


Η runExample() τρέχει όλους τους classifier

In [8]:
def runExample(someDf,testSize = 0.6):
    x = someDf.drop(['X65','ID'],axis = 1)
    x = StandardScaler().fit_transform(x)
    y = someDf['X65']
    X_train, x_test, y_train, y_test = \
    train_test_split(x,y,test_size=testSize)
    #train_size = (1 - testSize) %
    for name, clf in zip(names, classifiers):
            clf.fit(X_train, y_train)
            score = clf.score(x_test, y_test)
            predictions = clf.predict(x_test)
            print('\n\n\n #%s' % name)
            print(classification_report(y_test,predictions))

H compareScalers() τρέχει για κάθε scaler όλους τους classifier (Απίστευτα αργή συνάρτηση
έχει μπει σε σχόλιο και τα αποτελέσμα τις έχουν αποθηκευτεί σε αρχείο , πιο αναλυτικά πιο κάτω)

In [9]:
def compareScalers(tdf):
    scalerNames = ["Standard Scaler", "Mix max scaler",
         "Max abs Scaler", "Robust Scaler", "Quantile Transform uniform", "Quantile Transform normal",
         "Normalizer"]
    scalers = [
        StandardScaler(),
        MinMaxScaler(),
        MaxAbsScaler(),
        RobustScaler(quantile_range=(25, 75)),
        QuantileTransformer(output_distribution='uniform'),
        QuantileTransformer(output_distribution='normal'),
        Normalizer()
    ]
    for index,scaler in enumerate(scalers):
        print('>>>>>SCALER : %s  <<<<<<<' % scalerNames[index])
        someDf = tdf
        x = someDf.drop(['X65','ID'],axis = 1)
        x = scaler.fit_transform(x)
        y = someDf['X65']
        X_train, x_test, y_train, y_test = \
        train_test_split(x,y,test_size=0.6)
        #train_size = (1 - testSize) %
        for name, clf in zip(names, classifiers):
                clf.fit(X_train, y_train)
                score = clf.score(x_test, y_test)
                predictions = clf.predict(x_test)
                print('\n\n\n #%s and scaler %s' % (name,scalerNames[index]))
                print(classification_report(y_test,predictions))

Φόρτωση των αρχείων

In [10]:
from pathlib import Path
my_file = Path("/home/acid/Documents/ErgasiaDM/saved.df")

if my_file.is_file():
    df = pd.read_pickle("saved.df")
    print('loaded')
else:
    df = pd.read_csv('data/companydata.csv')
    df['ID'] = pd.Series((df.index), index= df.index)
    df = df.convert_objects(convert_numeric=True)
    df.to_pickle("saved.df")  
    print('created')
dfToSolve = pd.read_csv('toEvaluate/test_unlabeled.csv')
dfToSolve = dfToSolve.replace('?', 0.0)
dfToSolve = dfToSolve.convert_objects(convert_numeric=True)
dfToSolve = dfToSolve.fillna(df.mean())
dfToSolve.columns = ['X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X7', 'X8', 'X9','X10',
       'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X17', 'X18', 'X19',
       'X20', 'X21', 'X22', 'X23', 'X24', 'X25', 'X26', 'X27', 'X28',
       'X29', 'X30', 'X31', 'X32', 'X33', 'X34', 'X35', 'X36', 'X37',
       'X38', 'X39', 'X40', 'X41', 'X42', 'X43', 'X44', 'X45', 'X46',
       'X47', 'X48', 'X49', 'X50', 'X51', 'X52', 'X53', 'X54', 'X55',
       'X56', 'X57', 'X58', 'X59', 'X60', 'X61', 'X66', 'X63', 'X64'
       ]
from sklearn.utils import shuffle
df = shuffle(df)

loaded


/home/acid/Documents/ErgasiaDM/ErgEnv/lib/python2.7/site-packages/ipykernel_launcher.py:15: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  from ipykernel import kernelapp as app


Αφαιρούνται και στήλες με περισσότερα από 1000 na

In [ ]:
mistakeSeries = df.isnull().sum()
#Index starts at 0 and X starts at 1
indexCounter = 1
for index,item in enumerate(mistakeSeries):
    if(item > 1000 ):
        toRemove = 'X%s' % (index+indexCounter)
        df = df.drop(toRemove,axis = 1)
        index += 1 
        print(index)
# FIX THIS FIX THIS FIX THIS

21
37


# Τι θα γίνει με τις nan τιμές;
Συμπληρώνονται με βάση τον μέσω αριθμό τις στήλης , ένα απλό drop θα χαλούσε πολύ το std στις στείλες


In [ ]:
df = df.fillna(df.mean())
#first run
runExample(df)




 #NeuralNet
             precision    recall  f1-score   support

          0       0.96      1.00      0.98      4059
          1       1.00      0.01      0.01       158

avg / total       0.96      0.96      0.94      4217



# Test and decide on Scaler and classifier
## BEWARE This process takes a tone of time ! 
## uncomment comparedScalers(df) if you want to run it

In [ ]:
# output saved on ScalersResult.txt
#compareScalers(df) 
# best results saved as well

# The best results
## NeuralNet and scaler Standard Scaler
		     precision    recall  f1-score   support

		  0       0.96      1.00      0.98      4057
		  1       1.00      0.01      0.01       160

	avg / total       0.96      0.96      0.94      4217

## Linear SVM and scaler Quantile Transform normal
		     precision    recall  f1-score   support

		  0       0.96      1.00      0.98      4044
		  1       0.93      0.08      0.15       173

	avg / total       0.96      0.96      0.95      4217

## NeuralNet and scaler Quantile Transform normal
		     precision    recall  f1-score   support

		  0       0.97      1.00      0.98      4044
		  1       0.85      0.19      0.31       173

	avg / total       0.96      0.97      0.95      4217


## Decision Tree and scaler Quantile Transform normal
		     precision    recall  f1-score   support

		  0       0.98      1.00      0.99      4044
		  1       0.84      0.49      0.62       173

	avg / total       0.97      0.98      0.97      4217


## Gaussian Process and scaler Quantile Transform uniform
		     precision    recall  f1-score   support

		  0       0.98      1.00      0.99      4046
		  1       0.83      0.44      0.57       171

	avg / total       0.97      0.97      0.97      4217

## Decision Tree and scaler Quantile Transform uniform
		     precision    recall  f1-score   support

		  0       0.98      1.00      0.99      4046
		  1       0.81      0.45      0.58       171

	avg / total       0.97      0.97      0.97      4217



 ## Decision Tree and scaler Mix max scaler
		     precision    recall  f1-score   support

		  0       0.98      1.00      0.99      4064
		  1       0.80      0.42      0.56       153

	avg / total       0.97      0.98      0.97      4217

## AdaBoost and scaler Quantile Transform normal
		     precision    recall  f1-score   support

		  0       0.97      1.00      0.98      4044
		  1       0.80      0.35      0.49       173

	avg / total       0.97      0.97      0.96      4217

## Decision Tree and scaler Standard Scaler
		     precision    recall  f1-score   support

		  0       0.98      0.99      0.99      4057
		  1       0.74      0.50      0.60       160

	avg / total       0.97      0.97      0.97      4217

## Decision Tree and scaler Robust Scaler
		     precision    recall  f1-score   support

		  0       0.98      0.99      0.99      4062
		  1       0.70      0.45      0.54       155

	avg / total       0.97      0.97      0.97      4217


### We will use the Decision Tree and scaler Quantile Transform normal 
### best f1 score

dtQtn() runs on a dataframe and creates a new classifier
dtWithClassifier() παίρνει έναν ήδη υπάρχων classifier και τρέχει σε dataset

In [ ]:
def dtQtn(someDf,testSize = .6 ,maxDepth = 5,mute = 0):
    x = someDf.drop(['X65','ID'],axis = 1)
    x = QuantileTransformer(output_distribution='normal').fit_transform(x)
    y = someDf['X65']
    X_train, x_test, y_train, y_test = \
    train_test_split(x,y,test_size=testSize)
    #train_size = (1 - testSize) %
    clf = DecisionTreeClassifier(max_depth= maxDepth)
    clf.fit(X_train, y_train)
    score = clf.score(x_test, y_test)
    predictions = clf.predict(x_test)
    if(mute == 0):
        print(classification_report(y_test,predictions))
    return [clf,classification_report(y_test,predictions)]

def dtWithClassifier(someDf,classifier,testSize = .6,mute = 0):
    x = someDf.drop(['X65','ID'],axis = 1)
    x = QuantileTransformer(output_distribution='normal').fit_transform(x)
    y = someDf['X65']
    X_train, x_test, y_train, y_test = \
    train_test_split(x,y,test_size= testSize)
    clf = classifier
    clf.fit(X_train, y_train)
    score = clf.score(x_test, y_test)
    predictions = clf.predict(x_test)
    if(mute == 0):
        print(classification_report(y_test,predictions))
    return [clf,classification_report(y_test,predictions)]

In [ ]:
for i in [1,2,3,4,5,6,7,8,9,10]:
    print(i)
    dtQtn(df,maxDepth = i)
#Μετά από πολλές εκτελέσεις αποφασίσαμε ότι 4,6,7,9 είναι οι καλύτερες τιμέ για maxDepth

Τρέχουμε επαναλαμβανόμενα για να πάρουμε τον classifier με το καλύτερο f1score για την class 1
(αποφασίζεται δυναμικά το max depth)

In [ ]:
maxScore = 0
bestReport = None
classifier = None 
for i in range(0, 20):
    for i in [4,5,6,7,9]:
        result = dtQtn(df,testSize = .5,maxDepth = i,mute = 1)
        f1Score =  result[1].split('       ')[5].split('      ')[2]
        if (f1Score > maxScore):
            maxScore = f1Score
            classifier = result[0]
            bestReport = result[1]
print(maxScore)
print(bestReport)
# we got the best possible

In [ ]:
dtWithClassifier(df,classifier,mute = 0)[0]

# DataCleaning με pearson

In [ ]:
#trexoume gia kathe stili pearson kai rixnoume analoga
bankrupcy = df['X65']
toRunPearson = df.drop(['ID','X65'],axis =1)
pCleanUp = df
pearson = []
for item in toRunPearson:
    #print(bankrupcy)
    pearsonResult = pearsonr(toRunPearson[item],bankrupcy)[0]
    pearson.append([pearsonResult,item])
    if (pearsonResult < 0):
        pCleanUp = pCleanUp.drop([item],axis = 1)
        print('drop %s' % item )
apDf = pCleanUp



# After pearson

In [ ]:
#οι στήλες που μείναν
print(apDf.columns)
#Αυτές είναι οι στήλες που θα διαλέγαμε ως υποσύνολο για να αξιολογήσουμε , αλλά οι αφαίρεση στηλών (των υπόλοιπων)
#επηρέασε πολύ αρνητικά το μοντέλο ώστε να μπορούμε να εξάγουμε την αξία των γνωρισμάτων για το dataframe

In [ ]:
print('After pearson')
print(dtQtn(apDf,mute = 1)[1]) #after pearson
print('Before pearson')
print(dtWithClassifier(df,classifier,mute = 1)[1]) #before pearson


Τελικά δεν θα χρησιμοποιηθεί γιατί έχει πολύ αρνητικά αποτελέσματα

Το dataframe προς επίλυση

In [ ]:
copy = dfToSolve
copy = copy.fillna(copy.mean())
copy = copy[df.drop(['X65','ID'],axis = 1).columns]
#Drop the missing columns

# Πολύ σημαντικό! 
### Εφαρμόζουμε τον ίδιο scaler

In [ ]:
index = 0
solvedList = []
x = copy
x = QuantileTransformer(output_distribution='normal').fit_transform(x)

for item in classifier.predict_proba(copy):
    #Ελπίζουμε ότι τα ID δεν ανακατεύτηκαν 
    solvedList.append([index,item[0],item[1]])
    index +=1


In [ ]:
qSolved = pd.DataFrame(solvedList)
qSolved.columns = ['ID','NotBankruptProb','BankruptProb']
asked = qSolved.sort_values(by=['BankruptProb'],ascending = False)
ids = asked['ID'].head(50)

In [ ]:
#ΓΙΑΤΙ ΤΟ ROWID ΔΕΝ ΑΡΧΙΖΕΙ ΑΠΟ 0 :'(
idList = ids.tolist()
arraysStartAt0 = []
for item in idList:
    plzDontStartAt1 = item + 1
    arraysStartAt0.append(plzDontStartAt1)

In [ ]:
#Τα αποτελέσματα σε αρχείο
thefile = open('results.txt', 'w')
for index,item in enumerate(arraysStartAt0):
    if (index != 49):
        thefile.write("%s,\n" % item)
    else:
        thefile.write("%s" % item)
